validator/delphi_validator/validate.py

```diff
@@ -0,0 +1,742 @@
+                "date of most recent generated file seems too recent"))
+
+    def check_max_date_vs_reference(self, df_to_test, df_to_reference, checking_date, geo, sig):
+        """
+        Check if reference data is more recent than test data.
+
+        Arguments:
+            - df_to_test: pandas dataframe of a single CSV of source data
+            (one day-signal-geo_type combo)
+            - df_to_reference: pandas dataframe of reference data, either from the
+            COVIDcast API or semirecent data
+            - geo: str; geo type name (county, msa, hrr, state) as in the CSV name
+            - sig: str; signal name as in the CSV name
+
+        Returns:
+            - None
+        """
+        if df_to_test["time_value"].max() < df_to_reference["time_value"].max():
+            self.raised_errors.append(ValidationError(
+                ("check_max_date_vs_reference", checking_date.date(), geo, sig),
+                (df_to_test["time_value"].max(),
+                 df_to_reference["time_value"].max()),
+                'reference df has days beyond the max date in the =df_to_test=; ' +
+                'checks are not constructed to handle this case, and this situation ' +
+                'may indicate that something locally is out of date, or, if the local ' +
+                'working files have already been compared against the reference, ' +
+                'that there is a bug somewhere'))
+
+    def check_rapid_change_num_rows(self, df_to_test, df_to_reference, checking_date, geo, sig):
+        """
+        Compare number of obervations per day in test dataframe vs reference dataframe.
+
+        Arguments:
+            - df_to_test: pandas dataframe of CSV source data
+            - df_to_reference: pandas dataframe of reference data, either from the
+            COVIDcast API or semirecent data
+            - checking_date: datetime date
+            - geo: str; geo type name (county, msa, hrr, state) as in the CSV name
+            - sig: str; signal name as in the CSV name
+
+        Returns:
+            - None
+        """
+        test_rows_per_reporting_day = df_to_test[df_to_test['time_value']
+                                                 == checking_date].shape[0]
+        reference_rows_per_reporting_day = df_to_reference.shape[0] / len(
+            set(df_to_reference["time_value"]))
+
+        compare_rows = reldiff_by_min(
+            test_rows_per_reporting_day,
+            reference_rows_per_reporting_day)
+
+        if abs(compare_rows) > 0.35:
+            self.raised_errors.append(ValidationError(
+                ("check_rapid_change_num_rows", checking_date, geo, sig),
+                (test_rows_per_reporting_day, reference_rows_per_reporting_day),
+                "Number of rows per day (-with-any-rows) seems to have changed " +
+                "rapidly (reference vs test data)"))
+
+    def check_avg_val_diffs(self, df_to_test, df_to_reference, checking_date, geo, sig):
```

Comment:
This looks pretty complicated, but also seems like a technique that other people may be doing too. Do you know if there are there any anomaly detection/comparison packages we could take advantage of for this?

Reply:
>  are there any anomaly detection/comparison packages we could take advantage of for this?

Good thought! I'm aware of a few packages, but they are focused on modeling-based outlier detection, which might be a bit heavy-weight for our purposes here. I'll look into this and see if anything is promising.